airbyte-integrations/connectors/source-netsuite-odbc/source_netsuite_odbc/streams.py

```diff
@@ -0,0 +1,424 @@
+      stream_slice: Optional[Mapping[str, Any]] = None,
+      stream_state: Optional[Mapping[str, Any]] = None,
+    ) -> Iterable[StreamData]:
+      try:
+        self.process_stream_state(stream_state)
+        self.db_connection.execute(self.generate_query(sync_mode, stream_slice))
+        number_records = 0
+        while True:
+          row = self.db_connection.fetchone()
+          if not row and number_records < NETSUITE_PAGINATION_INTERVAL:
+            break
+          # if number of rows >= page limit, we need to fetch another page
+          elif not row:
+            self.logger.info(f"Fetching another page for the stream {self.table_name}.  Current state is (primary key: {str(self.primary_key_last_value_seen)}, date: {self.incremental_most_recent_value_seen}")
+            number_records = 0
+            self.db_connection.execute(self.generate_query(sync_mode, stream_slice))
+            continue
+          # data from netsuite does not include columns, so we need to assign each value to the correct column name
+          serialized_data = self.serialize_row_to_response(row)
+          # before we yield record, update state
+          self.update_last_values_seen(serialized_data, sync_mode)
+          self.find_most_recent_date(serialized_data)
+          number_records = number_records + 1
+          #yield response
+          yield serialized_data
+      except Exception as e:
+        message = str(e)
+        if NETSUITE_CONNECTION_EXPIRED_FAILURE in message:
+          self.logger.info(f"Connection expired for stream {self.table_name}.  Reconnecting...")
+          cursor_constructor = NetsuiteODBCCursorConstructor()
+          new_db_connection = cursor_constructor.create_database_connection(self.config)
+          self.db_connection = new_db_connection
+          current_stream_state = self.get_updated_state(stream_state, {})
+          self.read_records(sync_mode=sync_mode, cursor_field=cursor_field, stream_slice=stream_slice, stream_state=current_stream_state)
+        else:
+          raise e
+      self.logger.info(f"Finished Stream Slice for Netsuite ODBC Table {self.table_name} with stream slice: {stream_slice}")
+      #  Because the WHERE filter changes in each stream slice, primary key values are not guaranteed to be in order
+      #  across stream slives.  Therefore, we reset the primary key last value seen after each stream slice
+      if self.incremental_column is not None: #  if an incremental column does not exist, we do NOT reset the primary key
+        self.primary_key_last_value_seen = self.set_up_primary_key_last_value_seen(self.primary_key_column)
+      self.db_connection.execute("SELECT COUNT(*) FROM " + self.table_name)
+      self.logger.info(f"Finished reading stream slice for {self.table_name}.  Total number of records in table: {self.db_connection.fetchone()[0]}")
+
+    #  STATE MANAGEMENT FUNCTIONS
+    def process_stream_state(self, stream_state):
```

Comment:
nit: let's call this process_input_stream_state so we're clear this is only meant to be used for initializing the stream